decoder/src/log/stdout_logger.rs

```diff
@@ -131,50 +131,70 @@ impl StdoutLogger {
 }
 
 /// Printer for `DefmtRecord`s.
-struct Printer<'a> {
+pub(super) struct Printer<'a> {
     record: Record<'a>,
     format: &'a [LogSegment],
     min_timestamp_width: usize,
 }
 
 impl<'a> Printer<'a> {
-    pub fn new(record: Record<'a>, format: &'a [LogSegment]) -> Self {
+    fn new(record: Record<'a>, format: &'a [LogSegment]) -> Self {
         Self {
             record,
             format,
             min_timestamp_width: 0,
         }
     }
 
+    pub fn new_defmt(record: &'a DefmtRecord<'a>, format: &'a [LogSegment]) -> Self {
+        Self {
+            record: Record::Defmt(record),
+            format,
+            min_timestamp_width: 0,
+        }
+    }
+
     /// Pads the defmt timestamp to take up at least the given number of characters.
     /// TODO: Remove this, shouldn't be needed now that we have width field support
-    pub fn min_timestamp_width(&mut self, min_timestamp_width: usize) -> &mut Self {
+    fn min_timestamp_width(&mut self, min_timestamp_width: usize) -> &mut Self {
         self.min_timestamp_width = min_timestamp_width;
         self
     }
 
     /// Prints the formatted log frame to `sink`.
     pub fn print_frame<W: io::Write>(&self, sink: &mut W) -> io::Result<()> {
         for segment in self.format {
-            let s = match &segment.metadata {
-                LogMetadata::String(s) => s.to_string(),
-                LogMetadata::Timestamp => self.build_timestamp(&segment.format),
-                LogMetadata::FileName => self.build_file_name(&segment.format),
-                LogMetadata::FilePath => self.build_file_path(&segment.format),
-                LogMetadata::ModulePath => self.build_module_path(&segment.format),
-                LogMetadata::LineNumber => self.build_line_number(&segment.format),
-                LogMetadata::LogLevel => self.build_log_level(&segment.format),
-                LogMetadata::Log => self.build_log(&segment.format),
-                LogMetadata::NestedLogSegments(segments) => {
-                    self.build_nested(segments, &segment.format)
-                }
-            };
-
+            let s = self.build_segment(segment);
             write!(sink, "{s}")?;
         }
         writeln!(sink)
     }
 
+    pub(super) fn format_frame(&self) -> Result<String, std::fmt::Error> {
```

Comment:
I was thinking about that. But the Printer holds a reference to the record aka. the log frame. So you basically have one Printer per log frame, which makes sense because it uses this reference in almost all methods.

The Formatter should not keep a reference to the record, but just receive it for the method call.

Reply:
The name `Printer` does not really make sense anymore. I think `SingleRecordFormatter` would be the most correct, but maybe a bit bulky. But I'd like to pull the naming discussion out of this PR.

Reply:
As I have written as a `TODO` in code, I think the `StdoutLogger` should use two `Formatters` instead of just storing two log_formats. So the hierachy would be:

```mermaid
graph LR
    StdoutLogger --> Formatter
    Formatter --> Printer
```

Reply:
> The name Printer does not really make sense anymore. I think SingleRecordFormatter would be the most correct, but maybe a bit bulky. But I'd like to pull the naming discussion out of this PR.

I disagree, the purpose of `Printer` after this PR should not be to format, but to print. The only reason why `SingleRecordFormatter` would make sense after this PR, is because you added `format_frame` to `Printer`. I think `format_frame` should be the responsibility of `Formatter`.


> The Formatter should not keep a reference to the record, but just receive it for the method call.

I agree, then in `print_frame` the `Printer` would just need to call
```
write!(sink, "{}", self.formatter.format_frame(&self.record)):
```


>  I think the StdoutLogger should use two Formatters instead of just storing two log_formats.

Yeah, I think that makes sense.